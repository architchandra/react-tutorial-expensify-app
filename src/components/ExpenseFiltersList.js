import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters';



class ExpenseFiltersList extends React.Component {
  state = {
    calendarFocused: null,
  };

  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };
  onFocusChange = (focusedInput) => {
    this.setState(() => ({ calendarFocused: focusedInput }));
  };
  onTextChange = (e) => {
    this.props.setTextFilter(e.target.value);
  };
  onSortChange = (e) => {
    const sortByValue = e.target.value;

    if(sortByValue === 'date') {
      this.props.sortByDate();
    } else if(sortByValue === 'amount') {
      this.props.sortByAmount();
    }
  };

  render() {
    return (
      <div className="content-container">
        <div className="input-group">
          <div className="input-group__item">
            <input
              className="text-input"
              type="text"
              value={this.props.filters.text}
              onChange={this.onTextChange}
              placeholder="Search expenses"
            />
          </div>
          <div className="input-group__item">
            <select
              className="select"
              value={this.props.filters.sortBy}
              onChange={this.onSortChange}
            >
              <option value="date">Date</option>
              <option value="amount">Amount</option>
            </select>
          </div>
          <div className="input-group__item">
            <DateRangePicker
              startDate={this.props.filters.startDate}
              endDate={this.props.filters.endDate}
              onDatesChange={this.onDatesChange}
              focusedInput={this.state.calendarFocused}
              onFocusChange={this.onFocusChange}
              numberOfMonths={1}
              isOutsideRange={() => false}
              showClearDates={true}
            />
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  filters: state.filters,
});
const mapDispatchToProps = (dispatch) => ({
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount()),
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndDate: (endDate) => dispatch(setEndDate(endDate)),
});

export { ExpenseFiltersList };
export default connect(mapStateToProps, mapDispatchToProps)(ExpenseFiltersList);
