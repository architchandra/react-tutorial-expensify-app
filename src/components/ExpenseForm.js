import React from 'react';



class ExpenseForm extends React.Component {
  state = {
    description: '',
    amount: '',
    note: '',
  }

  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }))
  };

  onAmountChange = (e) => {
    const amount = e.target.value;
    if(amount.match(/^\d*(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };

  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };
  
  render() {
    return (
      <div>
        <form>
          <input
            type="text"
            value={this.state.description}
            placeholder="Description"
            onChange={this.onDescriptionChange}
            autoFocus
          />
          <input
            type="text"
            value={this.state.amount}
            placeholder="0.00"
            onChange={this.onAmountChange}
          />
          <textarea
            value={this.state.note}
            placeholder="Note"
            onChange={this.onNoteChange}
            rows="4"
          ></textarea>
        </form>
      </div>
    );
  }
}
export default ExpenseForm;