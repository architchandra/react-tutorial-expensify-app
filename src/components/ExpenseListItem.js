import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';



const ExpenseListItem = ({id, description, note, amount, createdAt}) => {
  return (
  <Link className="list-item" to={`/edit/${id}`}>
    <div>
      {description && <h3 className="list-item__title">{description}</h3>}
      {(createdAt !== null && (typeof createdAt !== 'undefined')) && <span className="list-item__subtitle">Added on {moment(createdAt).format('Do MMMM YYYY')}</span>}
    </div>
    {amount && <h3 className="list-item__data">{numeral(amount).format('$0,0.00')}</h3>}
  </Link>
)};
export default ExpenseListItem;
