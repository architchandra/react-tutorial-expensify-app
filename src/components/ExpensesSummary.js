import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import numeral from 'numeral';
import getExpenses from '../selectors/expenses';
import getExpensesTotal from '../selectors/expenses-total';



const ExpensesSummary = (props) => {
  const hasExpenses = props.expenses.length;
  const hasOnlyOneExpense = props.expenses.length === 1;
  
  return (
    <div className="page-header">
      <div className="content-container">
        <h2 className="page-header__title">
          <span>{props.expenses.length + (hasOnlyOneExpense ? ' expense' : ' expenses')}</span> found, totalling <span>{numeral(getExpensesTotal(props.expenses)).format('$0,0.00')}</span>.
        </h2>
        <div className="page-header__actions">
          <Link className="button" to="/create">
            Add Expense
          </Link>
        </div>
      </div>
    </div>
  );
};
const mapStateToProps = (state) => ({
  expenses: getExpenses(state.expenses, state.filters),
});



export { ExpensesSummary };
export default connect(mapStateToProps)(ExpensesSummary);
