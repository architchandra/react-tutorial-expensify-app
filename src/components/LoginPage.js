import React from 'react';
import { connect } from 'react-redux';
import { startLogin } from '../actions/auth';



const LoginPage = (props) => (
  <div className="box-layout">
    <div className="box-layout__box">
      <h1 className="box-layout__title">Expensify</h1>
      <p>Get your expenses under control</p>
      <button className="button" onClick={props.startLogin}>
        Login with Google
      </button>
    </div>
  </div>
);

const mapDispatchToProps = (dispatch) => ({
  startLogin: () => dispatch(startLogin()),
});



export { LoginPage };
export default connect(undefined, mapDispatchToProps)(LoginPage);
