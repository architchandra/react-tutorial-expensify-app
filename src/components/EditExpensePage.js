import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense } from '../actions/expenses';



const EditExpensePage = (props) => {
  return (
    <div>
      <h2>Edit an Expense</h2>
      <ExpenseForm
        expense={props.expense}
        onSubmit={(expense) => {
          props.dispatch(editExpense(props.expense.id, expense));
          props.history.push('/');
        }}
      />
    </div>
  );
};
const mapStateToProps = (state, props) => ({
  expense: state.expenses.find((expense) => expense.id === props.match.params.id),
});
export default connect(mapStateToProps)(EditExpensePage);