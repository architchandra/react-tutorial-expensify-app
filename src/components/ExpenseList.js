import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem.js';
import getExpenses from '../selectors/expenses.js';



const ExpenseList = (props) => (
  <div>
    <ul>
      {
        props.expenses.length !== 0
          ? props.expenses.map((expense, index) => <ExpenseListItem key={expense.id} {...expense} />)
          : <p>No expenses found.</p>
      }
    </ul>
  </div>
);
const mapStateToProps = (state) => ({
  expenses: getExpenses(state.expenses, state.filters)
});
export { ExpenseList };
export default connect(mapStateToProps)(ExpenseList);
