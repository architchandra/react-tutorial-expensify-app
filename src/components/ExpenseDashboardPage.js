import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseFiltersList from './ExpenseFiltersList';
import ExpensesSummary from './ExpensesSummary';



const ExpenseDashboardPage = () => (
  <div>
    <h2>Expense Dashboard Page</h2>
    <ExpensesSummary />
    <ExpenseFiltersList />
    <ExpenseList />
  </div>
);
export default ExpenseDashboardPage;
