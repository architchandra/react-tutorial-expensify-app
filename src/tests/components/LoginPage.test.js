import React from 'react';
import { shallow } from 'enzyme';
import { LoginPage } from '../../components/LoginPage';



let startLoginSpy, wrapper;
beforeEach(() => {
  startLoginSpy = jest.fn();
  wrapper = shallow(<LoginPage startLogin={startLoginSpy} />);
});



test('should render Login page', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should call start login on button click', () => {
  wrapper.find('button').simulate('click');
  expect(startLoginSpy).toHaveBeenCalled();
});
