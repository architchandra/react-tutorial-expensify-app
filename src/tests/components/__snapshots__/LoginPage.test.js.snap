// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Login page 1`] = `
<div>
  <button>
    Login
  </button>
</div>
`;
